File: android/app/src/main/java/org/switchboard/android/irc/IrcUrl.kt
package org.switchboard.android.irc

import java.net.URLDecoder

/**
 * What an `irc://` link means — the Kotlin half of `src/shared/ircurl.ts`,
 * checked against `tests/fixtures/ircurl.json`.
 */
object IrcUrl {

    data class Link(
        val host: String,
        val port: Int,
        val tls: Boolean,
        /** With its `#`, or null for a link to the server alone */
        val channel: String?,
        /** A person, when the link said `,isnick` */
        val nick: String?
    )

    // A `#` here is a channel, not a fragment — the one place on the web it is not
    private val SHAPE = Regex("^(ircs?)://([^/?#\\s]+)(?:/([^?\\s]*))?(?:\\?[^\\s]*)?$", RegexOption.IGNORE_CASE)
    private val HOST_PORT = Regex("^(\\[[^\\]]+\\]|[^:]+)(?::(\\d+))?$")

    fun parse(raw: String): Link? {
        val match = SHAPE.find(raw.trim()) ?: return null
        val tls = match.groupValues[1].lowercase() == "ircs"
        val authority = match.groupValues[2].substringAfter('@')
        val hostPort = HOST_PORT.find(authority) ?: return null
        val host = hostPort.groupValues[1].removePrefix("[").removeSuffix("]")
        val port = hostPort.groupValues[2].toIntOrNull() ?: if (tls) 6697 else 6667
        if (host.isEmpty() || port !in 1..65535) return null

        val path = match.groupValues[3]
        val parts = path.split(',')
        var target = URLDecoder.decode(parts[0], "UTF-8")
        val modifiers = parts.drop(1).map { it.lowercase() }
        if (target.isEmpty()) return Link(host, port, tls, null, null)
        if ("isnick" in modifiers) return Link(host, port, tls, null, target)
        if (target.first() !in "#&!+") target = "#$target"
        return Link(host, port, tls, target, null)
    }
}

File: src/shared/ircurl.ts
/**
 * What an `irc://` link means.
 *
 * `irc://irc.example.org/#channel`, `ircs://host:6697/channel` and the rest
 * of the forms the draft standard and twenty years of web pages have used.
 * Every desktop client registers for these; a link on a project's page
 * opened nothing here. One parser for both clients — the Kotlin half is
 * `IrcUrl.kt`, checked against `tests/fixtures/ircurl.json`.
 */

export interface IrcLink {
  host: string
  port: number
  tls: boolean
  /** With its `#`, or null for a link to the server alone */
  channel: string | null
  /** A person, when the link said `,isnick` */
  nick: string | null
}

export function parseIrcUrl(raw: string): IrcLink | null {
  // A `#` here is a channel, not a fragment — the one place on the web it is not
  const match = /^(ircs?):\/\/([^/?#\s]+)(?:\/([^?\s]*))?(?:\?[^\s]*)?$/i.exec(raw.trim())
  if (!match) return null
  const tls = match[1].toLowerCase() === 'ircs'
  const authority = match[2].replace(/^[^@]*@/, '')
  const hostPort = /^(\[[^\]]+\]|[^:]+)(?::(\d+))?$/.exec(authority)
  if (!hostPort) return null
  const host = hostPort[1].replace(/^\[|\]$/g, '')
  const port = hostPort[2] ? Number(hostPort[2]) : tls ? 6697 : 6667
  if (!host || !Number.isInteger(port) || port <= 0 || port > 65535) return null

  let target = decodeURIComponent(match[3] ?? '').split(',')[0]
  const modifiers = (match[3] ?? '').split(',').slice(1).map((m) => m.toLowerCase())
  if (target === '') return { host, port, tls, channel: null, nick: null }
  if (modifiers.includes('isnick')) return { host, port, tls, channel: null, nick: target }
  if (!/^[#&!+]/.test(target)) target = '#' + target
  return { host, port, tls, channel: target, nick: null }
}
